### DISCUSSION TITLE
Expanding plotting options of plot_posterior_predictive

### DISCUSSION THREAD

**Discussion by jsnyde0** [pymc-labs/pymc-marketing#947] [Ideas]
Hey! The plot_posterior_predictive has fairly limited options.

In order to see e.g. the posterior mean as well, I wrapped it in my own function that adds more to the returned fig.

I've never contributed to anyone else's code base, but would you be open to a PR where I add some plotting options like that?

**Comment by williambdean** [comment]
I think that's a good idea. I would love to see the plotting functionality have a bit more flexibility and extensibility. 

What changes to do you have in mind? 
There are some issues here: https://github.com/pymc-labs/pymc-marketing/issues?q=is%3Aopen+label%3Aplots+label%3AMMM
Feel free to add any more if you feel like

**Reply by jsnyde0** [reply]
Some simple things that I could do straight away:
* option to plot the posterior mean
* pass a list of HDI intervals you want to see

Something for which we'd have to think of how to organize it, would be an option to plot the posterior timeseries as a gradient (got this from a notebook by @juanitorduz, I believe):
![image](https://github.com/user-attachments/assets/fbcfab14-f4a9-424f-98c8-32356103cf87)

Perhaps I should just get started with the simple things to go through the process of contributing here, and when that works out work on a proposal for the gradient plot?

**Reply by williambdean** [reply]
I have to look at all the existing methods again, but is it possible to design where we can pass the fig into next plot? 

```python
# pseudo API 
fig, axes = mmm.plot_posterior_predictive_hdi(94)
mmm.plot_posterior_predictive_hdi(50, axes=axes)
mmm.plot_posterior_predictive_mean(axes=axes)
```

I just shy away from having methods with so so many arguments

**Reply by jsnyde0** [reply]
Ok how do you feel about an approach that allows both simple and more advanced usage? We can leave out the function arguments from `plot_posterior_predictive` if that's your preference.

## API
    def create_posterior_predictive_plot(ax=None, figsize=(10, 6), **plt_kwargs) -> plt.Axes:
        """Create base posterior predictive plot."""

    def add_hdi_to_plot(ax, hdi_prob, color=None, alpha=0.3, **kwargs) -> plt.Axes:
        """Add HDI to existing plot."""

    def add_mean_to_plot(ax, color='red', linestyle='-', **kwargs) -> plt.Axes:
        """Add mean prediction to existing plot."""

    def add_gradient_to_plot(ax, cmap='Blues', n_percentiles=100, **kwargs) -> plt.Axes:
        """Add gradient representation to existing plot."""

    def plot_posterior_predictive(hdi_probs=[0.94, 0.5], show_mean=True, 
                                  show_gradient=False, ax=None, **plt_kwargs) -> plt.Axes:
        """Convenience method for common plot configurations."""


## Simple usage
ax = mmm.plot_posterior_predictive()
plt.show()

## Advanced usage
fig, ax = plt.subplots(figsize=(15, 10))
ax = mmm.create_posterior_predictive_plot(ax)
ax = mmm.add_hdi_to_plot(ax, 0.94, color='blue', alpha=0.2)
ax = mmm.add_mean_to_plot(ax, color='red', linestyle='--')
plt.show()

**Reply by juanitorduz** [reply]
Hey @jsnyde0, Thanks for the input! What about extending the current function for backward compatibility? We can have individual private functions 

```python
def _add_hdi_to_plot(ax, hdi_prob, color=None, alpha=0.3, **kwargs) -> plt.Axes:
    """Add HDI to existing plot."""

def _add_mean_to_plot(ax, color='red', linestyle='-', **kwargs) -> plt.Axes:
    """Add mean prediction to existing plot."""

def _add_gradient_to_plot(ax, cmap='Blues', n_percentiles=100, **kwargs) -> plt.Axes:
    """Add gradient representation to existing plot."""
```
and then we allow this to happen in the `plot_posterior_predictive` as for example

```python
mmm.plot_posterior_predictive( . . . , add_mean: bool, ....):
    . . .
    if add_mean:
        _add_mean_to_plot( . . . )
    . . .
```

WDYT?

**Reply by jsnyde0** [reply]
Hi @juanitorduz, sounds fine with me. Will seemed to indicate he rather avoids methods with many arguments, which led to my previous proposal.

Going on vacation for a week but I can get started on a draft.

Thanks for your input!

**Reply by juanitorduz** [reply]
Sure! Take your time! ☀️ 🌴 !

**Reply by jsnyde0** [reply]
I've implemented this and created an issue for it: [#1057 Expanding plotting options of plot_posterior_predictive](https://github.com/pymc-labs/pymc-marketing/issues/1057).

Will continue discussion there, I guess?
